File: v2_typescript/src/portfolio_state.ts
const ROUND = 4;

function roundF64(v: number): number {
  const m = Math.pow(10, ROUND);
  return Math.round(v * m) / m;
}

function normOutcome(outcome: string): "UP" | "DOWN" | null {
  const o = outcome.trim().toUpperCase();
  if (o === "YES" || o === "UP") return "UP";
  if (o === "NO" || o === "DOWN") return "DOWN";
  return null;
}

export interface Position {
  shares: number;
  cost: number;
}

export interface OpenOrder {
  conditionId: string;
  outcome: string;
  side: string;
  size: number;
  price: number;
}

export class PortfolioState {
  cashBalance = 0;
  realizedPnl = 0;
  positions: Map<string, Map<string, Position>> = new Map();
  openOrders: Map<string, OpenOrder> = new Map();
  private applied: Map<string, number> = new Map();

  constructor(startingCash: number) {
    this.cashBalance = roundF64(startingCash);
  }

  private market(cid: string): Map<string, Position> {
    let m = this.positions.get(cid);
    if (!m) {
      m = new Map([
        ["UP", { shares: 0, cost: 0 }],
        ["DOWN", { shares: 0, cost: 0 }],
      ]);
      this.positions.set(cid, m);
    }
    return m;
  }

  private applyFill(
    cid: string,
    outcome: string,
    side: string,
    size: number,
    price: number
  ): void {
    const out = normOutcome(outcome);
    if (!out) return;
    const sz = roundF64(size);
    const pr = roundF64(price);
    const value = roundF64(sz * pr);
    const s = side.trim().toUpperCase();
    const market = this.market(cid);
    const pos = market.get(out)!;
    if (s === "BUY") {
      pos.shares = roundF64(pos.shares + sz);
      pos.cost = roundF64(pos.cost + value);
      this.cashBalance = roundF64(this.cashBalance - value);
      return;
    }
    if (pos.shares <= 0) return;
    const sellSize = roundF64(Math.min(size, pos.shares));
    if (sellSize <= 0) return;
    const avg = pos.shares > 0 ? pos.cost / pos.shares : 0;
    const sellValue = roundF64(sellSize * price);
    pos.shares = roundF64(pos.shares - sellSize);
    pos.cost = roundF64(pos.cost - (avg * sellSize));
    if (pos.shares <= 0) {
      pos.shares = 0;
      pos.cost = 0;
    }
    this.cashBalance = roundF64(this.cashBalance + sellValue);
    this.realizedPnl = roundF64(this.realizedPnl + sellSize * (price - avg));
  }

  applyImmediateFill(
    orderId: string,
    conditionId: string,
    outcome: string,
    side: string,
    size: number,
    price: number
  ): void {
    const out = normOutcome(outcome);
    if (!out) return;
    const prev = this.applied.get(orderId) ?? 0;
    const delta = roundF64(size) - prev;
    if (delta <= 0) return;
    this.applyFill(conditionId, out, side, delta, price);
    this.applied.set(orderId, roundF64(size));
  }

  onOrderUpdate(
    orderId: string,
    sizeMatched: number,
    conditionId: string,
    outcome: string,
    side: string,
    price: number
  ): boolean {
    const prev = this.applied.get(orderId) ?? 0;
    const delta = roundF64(sizeMatched - prev);
    if (delta <= 0) return true;
    const out = normOutcome(outcome);
    if (!out) return false;
    this.applyFill(conditionId, out, side, delta, price);
    this.applied.set(orderId, sizeMatched);
    return true;
  }

  registerOrder(
    orderId: string,
    conditionId: string,
    outcome: string,
    side: string,
    size: number,
    price: number
  ): void {
    const outcomeStr =
      outcome.trim().toUpperCase().replace("YES", "UP").replace("NO", "DOWN") ||
      "UP";
    const sideStr = side.trim().toUpperCase() || "BUY";
    this.openOrders.set(orderId, {
      conditionId,
      outcome: outcomeStr,
      side: sideStr,
      size,
      price,
    });
  }

  unregisterOrder(orderId: string): void {
    this.openOrders.delete(orderId);
  }

  getPosition(conditionId: string): [number, number, number, number] {
    const m = this.positions.get(conditionId);
    if (!m) return [0, 0, 0, 0];
    const u = m.get("UP") ?? { shares: 0, cost: 0 };
    const d = m.get("DOWN") ?? { shares: 0, cost: 0 };
    return [u.shares, d.shares, u.cost, d.cost];
  }
}

File: v2_rust/src/portfolio_state.rs
//! Portfolio: cash, positions per market (UP/DOWN), open orders. Same logic as v2_python utils/portfolio_state.

use std::collections::HashMap;
use std::sync::RwLock;

const ROUND: i32 = 4;

fn round_f64(v: f64) -> f64 {
    let m = 10_f64.powi(ROUND);
    (v * m).round() / m
}

fn norm_outcome(outcome: &str) -> Option<&'static str> {
    let o = outcome.trim().to_uppercase();
    if o == "YES" || o == "UP" {
        return Some("UP");
    }
    if o == "NO" || o == "DOWN" {
        return Some("DOWN");
    }
    None
}

#[derive(Clone, Debug, Default)]
pub struct Position {
    pub shares: f64,
    pub cost: f64,
}

fn empty_position() -> Position {
    Position {
        shares: 0.0,
        cost: 0.0,
    }
}

#[derive(Clone, Debug)]
pub struct OpenOrder {
    pub condition_id: String,
    pub outcome: String,
    pub side: String,
    pub size: f64,
    pub price: f64,
}

pub struct PortfolioState {
    pub cash_balance: RwLock<f64>,
    pub realized_pnl: RwLock<f64>,
    pub positions: RwLock<HashMap<String, HashMap<String, Position>>>, // cid -> UP/DOWN -> Position
    pub open_orders: RwLock<HashMap<String, OpenOrder>>,
    applied: RwLock<HashMap<String, f64>>, // order_id -> size already applied
}

impl PortfolioState {
    pub fn new(starting_cash: f64) -> Self {
        Self {
            cash_balance: RwLock::new(round_f64(starting_cash)),
            realized_pnl: RwLock::new(0.0),
            positions: RwLock::new(HashMap::new()),
            open_orders: RwLock::new(HashMap::new()),
            applied: RwLock::new(HashMap::new()),
        }
    }

    fn market(&self, cid: &str) -> HashMap<String, Position> {
        let mut positions = self.positions.write().unwrap();
        positions
            .entry(cid.to_string())
            .or_insert_with(|| {
                let mut m = HashMap::new();
                m.insert("UP".to_string(), empty_position());
                m.insert("DOWN".to_string(), empty_position());
                m
            })
            .clone()
    }

    fn apply_fill(
        &self,
        cid: &str,
        outcome: &str,
        side: &str,
        size: f64,
        price: f64,
    ) {
        let outcome = match norm_outcome(outcome) {
            Some(o) => o.to_string(),
            None => return,
        };
        let size = round_f64(size);
        let price = round_f64(price);
        let value = round_f64(size * price);
        let side = side.trim().to_uppercase();

        let mut positions = self.positions.write().unwrap();
        let market = positions
            .entry(cid.to_string())
            .or_insert_with(|| {
                let mut m = HashMap::new();
                m.insert("UP".to_string(), empty_position());
                m.insert("DOWN".to_string(), empty_position());
                m
            });
        let pos = market.get_mut(&outcome).unwrap();

        if side == "BUY" {
            let current_cash = *self.cash_balance.read().unwrap();
            pos.shares = round_f64(pos.shares + size);
            pos.cost = round_f64(pos.cost + value);
            drop(positions);
            *self.cash_balance.write().unwrap() = round_f64(current_cash - value);
            return;
        }
        // SELL
        if pos.shares <= 0.0 {
            return;
        }
        let sell_size = round_f64(size.min(pos.shares));
        if sell_size <= 0.0 {
            return;
        }
        let avg = if pos.shares > 0.0 {
            pos.cost / pos.shares
        } else {
            0.0
        };
        let sell_value = round_f64(sell_size * price);
        let current_cash = *self.cash_balance.read().unwrap();
        let current_pnl = *self.realized_pnl.read().unwrap();
        pos.shares = round_f64(pos.shares - sell_size);
        pos.cost = round_f64(pos.cost - avg * sell_size);
        if pos.shares <= 0.0 {
            pos.shares = 0.0;
            pos.cost = 0.0;
        }
        drop(positions);
        *self.cash_balance.write().unwrap() = round_f64(current_cash + sell_value);
        *self.realized_pnl.write().unwrap() =
            round_f64(current_pnl + sell_size * (price - avg));
    }

    /// Apply fill when place-order response is MATCHED.
    pub fn apply_immediate_fill(
        &self,
        order_id: &str,
        condition_id: &str,
        outcome: &str,
        side: &str,
        size: f64,
        price: f64,
    ) {
        let out = match norm_outcome(outcome) {
            Some(o) => o,
            None => return,
        };
        let mut applied = self.applied.write().unwrap();
        let prev = applied.get(order_id).copied().unwrap_or(0.0);
        let delta = round_f64(size) - prev;
        if delta <= 0.0 {
            return;
        }
        drop(applied);
        self.apply_fill(condition_id, out, side, delta, price);
        self.applied.write().unwrap().insert(order_id.to_string(), round_f64(size));
    }

    /// Apply incremental fill from user WSS UPDATE.
    pub fn on_order_update(
        &self,
        order_id: &str,
        size_matched: f64,
        condition_id: &str,
        outcome: &str,
        side: &str,
        price: f64,
    ) -> bool {
        let total = size_matched;
        let mut applied = self.applied.write().unwrap();
        let prev = applied.get(order_id).copied().unwrap_or(0.0);
        let delta = round_f64(total - prev);
        if delta <= 0.0 {
            return true;
        }
        let outcome = match norm_outcome(outcome) {
            Some(o) => o.to_string(),
            None => return false,
        };
        drop(applied);
        self.apply_fill(condition_id, &outcome, side, delta, price);
        self.applied.write().unwrap().insert(order_id.to_string(), total);
        true
    }

    pub fn register_order(
        &self,
        order_id: &str,
        condition_id: &str,
        outcome: &str,
        side: &str,
        size: f64,
        price: f64,
    ) {
        let outcome = (outcome.trim().to_uppercase())
            .replace("YES", "UP")
            .replace("NO", "DOWN");
        let outcome_str = if outcome.is_empty() { "UP".to_string() } else { outcome };
        let side = side.trim().to_uppercase();
        let side_str = if side.is_empty() { "BUY".to_string() } else { side };
        self.open_orders.write().unwrap().insert(
            order_id.to_string(),
            OpenOrder {
                condition_id: condition_id.to_string(),
                outcome: outcome_str,
                side: side_str,
                size,
                price,
            },
        );
    }

    pub fn unregister_order(&self, order_id: &str) {
        self.open_orders.write().unwrap().remove(order_id);
    }

    /// (qty_up, qty_down, cost_up, cost_down)
    pub fn get_position(&self, condition_id: &str) -> (f64, f64, f64, f64) {
        let positions = self.positions.read().unwrap();
        let market = match positions.get(condition_id) {
            Some(m) => m,
            None => return (0.0, 0.0, 0.0, 0.0),
        };
        let empty = empty_position();
        let u = market.get("UP").unwrap_or(&empty);
        let d = market.get("DOWN").unwrap_or(&empty);
        (u.shares, d.shares, u.cost, d.cost)
    }
}
